working-w-copilot/copilot-in-projects-refactoring.md

```diff
@@ -0,0 +1,559 @@
+    except:
+        abort(make_response({"message": f"{model_class.__name__.lower()} {model_id} invalid"}, 400))
+
+    model = model_class.query.get(model_id)
+
+    if not model:
+        abort(make_response({"message": f"{model_class.__name__.lower()} {model_id} not found"}, 404))
+
+    return model
+```
+
+Our function is updated, now we just need to update our dependent functions and tests to use `validate_model` instead of `validate_book`, which includes passing in the new `model_class` parameter. We can do this manually using VS Code's Find & Replace tools then adding in our new parameter, or we can reach out to Copilot. 
+
+One option is to open each file, select the whole file, and ask Copilot to change any instances of `validate_book` to `validate_model`. We need to carefully review the lines Copilot suggests changing to ensure nothing unexpected is altered, but it can save us a little time since we need to do slightly more than find and replace the function name.
+
+In our experimentation, we found that in the `routes.py` file, we could use a prompt with very few details to get the changes that we wanted: 
+
+> Please replace all uses of validate_book with validate_model
+
+However, in the `test_routes.py` file, we had to add more detail for Copilot to recognize that there was a new class parameter which was required wherever the function was invoked, and that we needed import the `Book` class:
+
+> Please replace all uses of validate_book(book_id) with validate_model(model_class, model_id). Import any required model classes.
+
+## Summary
+
+This hands-on experience in a project wraps up our whirlwind tour of GitHub's Copilot! Whether we're writing new code or working in an existing base Copilot can help us with tasks like generating ideas for tests, surfacing error handling we may have missed, and generally writing code quicker. Even if Copilot's suggestions look good at a glance, we always need to carefully proofread them both for correctness and completeness. Just because a code suggestion works does not mean it fits all of our needs, so we need to remain vigilant and critical of what Copilot provides. If you're interested in more practice with Copilot, we highly suggest revisiting Solar System, Task List, Inspo Board, or any other project that you're already familiar with so you know what the end product could look like and can focus on working with Copilot.
+
+## Check for Understanding
+
+<!-- Question 1 -->
+<!-- prettier-ignore-start -->
+### !challenge
+* type: checkbox
+* id: b53afd64-55e1-4670-8d3b-bce1ca3b8872
+* title: Refactoring with Copilot Concerns
+
+##### !question
+Select the refactoring steps below where Copilot may **not** be useful. 
+##### !end-question
+
+##### !options
+* Gathering requirements
+* Locating Dependencies
+* Checking and expanding our test suite
+* Executing the code change
+##### !end-options
+
+##### !answer
+* Gathering requirements
+* Locating Dependencies
+##### !end-answer
```

Comment:
- use labeled syntax
- executing could sound like running the code after we've made the change
```suggestion
##### !options
a| Gathering requirements
b| Locating Dependencies
c| Checking and expanding our test suite
d| Apply the code change
##### !end-options

##### !answer
a|
b|
##### !end-answer
```